### PR TITLE
Fix FP16 overflow in GQA attention and concat_past_present buffer overflow

### PR DESCRIPTION
**Pull request by aditya-dl**
## Motivation
Qwen and DeepSeek-Distill-Qwen models produce NaN/garbage output with MIGraphX FP16 inference. Large k_proj.bias values (~400) cause Q*K dot product results to reach ~301k, exceeding FP16 max (65504). This overflows to +inf, and softmax computes inf - inf = NaN.

## Technical Details
Insert convert(f16->f32) after the Q*K dot output by extending find_softmax_base_ops in rewrite_reduce.cpp to walk backwards from softmax through where/mul to find the upstream dot. The dot stays as dot(f16,f16)->f16 — the convert is placed on its output, extending the existing FP32 softmax upcast range to cover the entire dot→mul→where→softmax chain.

MFMA/WMMA hardware accumulates the dot product in FP32 internally. When the attention is fused, rocMLIR's RemoveRedundantCasts pass (rocMLIR PR https://github.com/ROCm/rocMLIR/pull/2202) eliminates the f32→f16→f32 roundtrip at LLVM IR level, preserving the FP32 accumulator value through to softmax.

Updated find_kv_cache_attention matcher in fuse_attention.cpp with skip(convert) so the matcher sees through the new convert to the dot. Attention fusion is preserved.

Changes:
`src/rewrite_reduce.cpp`: find_upstream_dot walks backwards from softmax to find upstream dot using unfold. Extended FP32 upcast range from [softmax_input, div] to [dot, div]. Skip bool and already-f32 inputs in upcast.
`src/fuse_attention.cpp`: skip(convert) between gemm1 and scale/attn_scores/local_window_mask/mask in find_kv_cache_attention matcher.
`test/rewrite_reduce.cpp`: 8 new tests: dot+scale+where, dot+scale, dot-only, no-dot-found, mul-no-upstream-dot, double-where, dot-relu (blocked), scale-on-left. Expected-module comparison pattern.
`test/fuse_attention.cpp`: End-to-end test running both rewrite_reduce and fuse_attention on a GQA graph, verifying fusion still produces group{kv_cache_attention}.
`test/verify/test_group_query_attention.cpp`: Regression test with large Q/K bias values (~400) triggering FP16 overflow.

## Dependency
This PR depends on the rocMLIR pin being updated to include `https://github.com/ROCm/rocMLIR/pull/2202 (merged April 2, 2026, commit 281add9)`. CI tests involving attention fusion will fail until the pin is updated.

## Changelog Category

Add a `CHANGELOG.md` entry for any option other than `Not Applicable`
- - [ ] Added: New functionality.
- - [ ] Changed: Changes to existing functionality.
- - [ ] Removed: Functionality or support that has been removed. (Compared to a previous release)
- - [ ] Optimized: Component performance that has been optimized or improved.
- - [ ] Resolved Issues: Known issues from a previous version that have been resolved.
- - [ ] Not Applicable: This PR is not to be included in the changelog.
-

### CODE REVIEW

**Comment by pfultz2** [comment]
Unit tests need to be added and the CI failure fixed.

File: src/include/migraphx/op/concat_past_present.hpp [hunk 1]
```diff
@@ -138,13 +147,21 @@ struct concat_past_present
...
         cache_params.seqlen_present_kv_cache = past_sequence_length;
 
-        visit_all(past, present)([&](auto past_kv, auto present_kv) {
+        // When output is larger than past cache (prompt case), allocate
+        // a properly-sized buffer so update_cache does not overflow.
+        argument output = past;
+        if (sequence_length > past_sequence_length)
+        {
+            output = argument{output_shape};
+        }
```
Comment (pfultz2) [inline src/include/migraphx/op/concat_past_present.hpp]: This makes the semantics of the operator confusing as it sometime aliases and sometimes it doesnt. We also are planning to remove this operator and replace it with `insert_slice`. This might complicate this transition. 

It also doesnt make sense to copy this to an internal buffer as that defeats the purpose of having a kv cache. 

Also, the way the kv cache currently works, It also doesnt make sense for the `past_sequence_length` to be larger then the `sequence_length` as the `past_sequence_length` should always be set to the max sequence length.

**Comment by pfultz2** [comment]
I think you should put the softmax change in a separate PR as I dont think we will merge the concat_past_present change.

File: src/include/migraphx/op/concat_past_present.hpp (hunk 1 above)
Comment (aditya-dl) [reply src/include/migraphx/op/concat_past_present.hpp]: @pfultz2 I've removed the concat_past_present update. This bug was introduced because of my testing script which was incorrectly creating a KV cache. Rectified.

**Comment by aditya-dl** [comment]
@pfultz2 Updated this PR with only the softmax change.

**Comment by aditya-dl** [comment]
The CI failures in `test_ck_gemm_softmax_gemm_0` and `test_ck_gemm_softmax_gemm_1` were caused by the backward walk unconditionally extending the FP32 upcast to the upstream dot. For non-masked attention patterns (dot→mul→softmax),  propagate_precision merges the FP32 converts into a clean chain, and fuse_attention fuses it into an MLIR attention module. rocMLIR's high-level pipeline then fails because it can't handle FP32 tosa.reduce_max inside fused attention.

The fix gates the backward walk on the presence of where in the chain. Without where, the dot stays in FP16 and CK fused attention handles precision internally. With where (all GQA models using the GroupQueryAttention ONNX op — Qwen, Llama, Phi, DeepSeek), the dot is upcast to FP32. The where ops block propagate_precision from merging converts (multi-input op), which prevents fuse_attention from matching, so ops run as separate FP32 kernels — fixing the FP16 overflow.

Added a new test (softmax_dot_no_where_preserves_half) verifying the dot is NOT upcast when where is absent.

File: src/rewrite_reduce.cpp [hunk 2]
```diff
@@ -95,20 +95,55 @@ struct find_softmax_base_ops
...
+                    found_where = true;
+                    current = current->inputs()[2]; // attention data is 3rd arg
+                }
+                else if(name == "mul" || name == "broadcast" ||
+                        name == "multibroadcast" || name == "convert")
+                    current = current->inputs()[0];
+                else
+                    break;
+            }
+        }
```
Comment (CharlieL7) [inline src/rewrite_reduce.cpp]: Would be cleaner to break this out into another function

File: src/rewrite_reduce.cpp [hunk 3]
```diff
@@ -95,20 +95,55 @@ struct find_softmax_base_ops
...
-        auto softmax_inss = find_instructions_between(inp, div, &m);
+        // Walk backwards from the softmax input through the attention chain
+        // (mul, where, broadcast, convert) to find a feeding dot instruction.
+        // This extends the FP32 upcast range to cover the entire dot->softmax
+        // chain, preventing FP16 overflow in attention score computation.
+        auto range_start = inp;
+        {
+            auto current = inp;
+            bool found_where = false;
+            for(int depth = 0; depth < 10; ++depth)
```
Comment (CharlieL7) [inline src/rewrite_reduce.cpp]: Is this depth arbitrary or is there an expected value?

**Review by CharlieL7 — CHANGES_REQUESTED** [review]

File: src/rewrite_reduce.cpp (hunk 2 above)
Comment (pfultz2) [inline src/rewrite_reduce.cpp]: This does a lot in one function and its hard to tell what it is doing. This can be written with `unfold`, `std::find_if` and `std::adjacent_find` to make it a lot cleaner:

```cpp
auto p = unfold(inp, [](instruction_ref x) -> std::optional<instruction_ref> {
    if(x->inputs().size() == 1)
        return x->inputs().front();
    if(contains({"add", "mul"}, x->name()))
    {
        auto it = std::find_if(x->inputs().begin(), x->inputs().end(), [](instruction_ref input) {
            return not input->can_eval();
        });
        if(it == x->inputs().end())
            return nullopt;
        return *it;
    }
    if(x->name() == "where")
        return x->inputs().at(2); // attention data is 3rd arg
    return nullopt;
});

auto where_it = std::find_if(p.begin(), p.end(), [&](instruction_ref x) {
    return x->name() == "where";
});
if(where_it != p.end())
{
    range_start = std::adjacent_find(where_it, p.end(), [&](instruction_ref, instruction_ref x) {
        return x->name() == "dot";
    });
}
```

File: test/rewrite_reduce.cpp [hunk 4]
```diff
@@ -78,6 +78,95 @@ TEST_CASE(softmax_upcast)
...
+    // All non-bool ops in the chain (mul, where, sub, exp, div) should be in FP32
+    auto chain_ops = find_all(migraphx::iterator_for(m), [](auto ins) {
+        auto name = ins->name();
+        return name == "mul" or name == "where" or name == "sub" or
+                name == "exp" or name == "div";
+    });
+    EXPECT(all_of(chain_ops, [](auto ins) {
+        return ins->get_shape().type() == migraphx::shape::float_type;
+    }));
+}
```
Comment (pfultz2) [inline test/rewrite_reduce.cpp:196]: Tests should be written with the expected module:

```cpp
module m1;
{
    // Create test module
}

run_pass(m1);

module m2;
{
    // Create expected module
}

EXPECT(m1 == m2);
```

**Review by pfultz2 — CHANGES_REQUESTED** [review]

**Comment by aditya-dl** [comment]
The found_where gate doesn't fully resolve the CI failure. 
`test_group_query_attention_decode_small` still fails because `propagate_precision` merges the FP32 converts for the small test shapes (even with where present), allowing `fuse_attention` to fuse the FP32 chain into an MLIR attention module. rocMLIR then fails on FP32 tosa.reduce_max ("can't trace the reduction output to a kernel result").

The core issue: `find_softmax_base_ops` intentionally upcasts the softmax chain to FP32, but `fuse_attention` later fuses it into an MLIR attention module that rocMLIR can't compile with FP32 reduce operations.

What's the recommended way to prevent fuse_attention from fusing when the softmax chain has been upcast to FP32? For example, would a type check in find_attention::apply() be appropriate?
```
  if(softmax_input->get_shape().type() == shape::float_type)
      return;
```

I'll also address the code review comments (unfold refactor, test restructure, depth parameter) once the CI issue is resolved.

**Comment by pfultz2** [comment]
> rocMLIR then fails on FP32 tosa.reduce_max ("can't trace the reduction output to a kernel result").

This doesnt look like data type error. It looks like the graph was changed in a way the rocMLIR can no longer trace it back to the gemm.

> What's the recommended way to prevent fuse_attention from fusing when the softmax chain has been upcast to FP32? 

We dont want to prevent attention fusion when fp32 softmax is used. We need to make sure this works by either fixing the graph we give to rocMLIR or updating the rocMLIR to support this case.

**Comment by aditya-dl** [comment]
> > rocMLIR then fails on FP32 tosa.reduce_max ("can't trace the reduction output to a kernel result").
> 
> This doesnt look like data type error. It looks like the graph was changed in a way the rocMLIR can no longer trace it back to the gemm.
> 
> > What's the recommended way to prevent fuse_attention from fusing when the softmax chain has been upcast to FP32?
> 
> We dont want to prevent attention fusion when fp32 softmax is used. We need to make sure this works by either fixing the graph we give to rocMLIR or updating the rocMLIR to support this case.

@pfultz2 can you advice on a way to debug this to identify the exact root cause of this? This doesn't replicate on my local system with RDNA GPUs.

**Comment by aditya-dl** [comment]
Updated this PR with a new approach based on feedback.

Previous approach: Upcasted dot inputs to FP32, making the entire dot→softmax chain run in FP32. This broke attention fusion because rocMLIR's getSoftmaxType() couldn't find the expected f16→f32 cast when GEMM1 was already f32.

New approach: Keep the dot as dot(f16,f16)→f16 and insert convert(f16→f32) after the dot output. This extends the existing FP32 softmax upcast range to cover mul/where between dot and softmax. The dot's graph pattern is preserved, so find_kv_cache_attention fusion still works (with skip(convert) added to the matcher).

The MFMA/WMMA hardware accumulates in f32 internally. rocMLIR's RemoveRedundantCasts pass (https://github.com/ROCm/rocMLIR/pull/2202, now merged) eliminates the f32→f16→f32 roundtrip at LLVM IR level, preserving the f32 accumulator value through to softmax.

Addressed reviewer feedback: refactored backward walk to use unfold, rewrote tests to compare expected vs actual modules, added edge case tests.

This PR depends on a rocMLIR pin update to include ROCm/rocMLIR#2202.

File: src/rewrite_reduce.cpp [hunk 5]
```diff
@@ -85,6 +86,38 @@ struct find_softmax_base_ops
 
     auto matcher() const { return match::softmax(); }
 
+    // Walk backwards from the softmax input through the attention chain
+    // (where, mul, broadcast, multibroadcast, convert) to find an upstream dot.
+    // Returns the dot instruction, or nullopt if not found.
```
Comment (shivadbhavsar) [inline src/rewrite_reduce.cpp]: I dont think this belongs in the "find_softmax_base_ops" finder. It would be better to let this transform happen, and then have a matcher that looks for the specific pattern you are trying to find here. Ie. `dot -> <elementwise ops> -> convert -> softmax`. You can then add the convert after the dot in the apply method of that finder. 

This way if we ever need to disable this behavior or have it toggle with a flag or anything, it is easy to decouple this part from the basic softmax upcast

Comment (aditya-dl) [reply src/rewrite_reduce.cpp]: @shivadbhavsar That makes sense, I've updated the code in the recent commit. Please review it.

File: src/rewrite_reduce.cpp [hunk 6]
```diff
@@ -79,6 +80,94 @@ struct find_softmax
...
+//
+// Runs before find_softmax_base_ops so that the softmax internals
+// (reduce_max through div) are still in f16 when find_softmax_base_ops
+// processes them.
+struct find_dot_softmax_fp32
+{
+    auto matcher() const { return match::softmax(); }
+
+    // Walk backwards from the softmax input through the attention chain
+    // (where, mul, broadcast, multibroadcast, convert) to find an upstream dot.
```
Comment (shivadbhavsar) [inline src/rewrite_reduce.cpp]: If this is the specific op chain you are looking for, it should be written as a composed matcher. 

But I think this can be better generalized, this sort of pattern matching should happen in the matcher when possible; apply method is for doing the transformation once that pattern is found. Traversing up from softmax in the matcher might be difficult due to non-unary ops but you can traverse down from the dot using outputs. 

Something like:
```
auto matcher() const
{
    auto attention_ops = match::name("mul", "where", "broadcast", "multibroadcast");
    return match::name("dot")(
        match::skip_output(attention_ops)(
            match::name("convert")(match::softmax_output()).bind("convert_ins")
        )
    );
}
```
and you can define softmax_output in softmax.hpp where match::softmax is defined with something like:
```
    auto div_out        = f("div");
    auto reduce_sum_out = f("reduce_sum")(
                              skip_output(f("multibroadcast", "broadcast"))(div_out));
    auto exp_out        = f("exp")(
                              any_of[outputs()](div_out),
                              any_of[outputs()](reduce_sum_out));
    auto sub_out        = f("sub")(any_of[outputs()](exp_out));
    auto reduce_max_out = f("reduce_max")(
                              skip_output(f("multibroadcast", "broadcast"))(sub_out));
    return any_of[outputs()](reduce_max_out);
```

This makes the op chain between the dot and softmax easily extendible too

File: src/rewrite_reduce.cpp [hunk 7]
```diff
@@ -218,6 +307,7 @@ void rewrite_reduce::apply(module& m) const
 
     if(not enabled(MIGRAPHX_DISABLE_FP32_SOFTMAX{}))
     {
+        match::find_matches(m, find_dot_softmax_fp32{});
         match::find_matches(m, find_softmax_base_ops{});
```
Comment (shivadbhavsar) [inline src/rewrite_reduce.cpp:307]: The new matcher should come after the base_ops transformation

**Review by shivadbhavsar — CHANGES_REQUESTED** [review]

**Comment by aditya-dl** [comment]
@shivadbhavsar Thanks for the feedback on using a composed matcher, that's a cleaner approach than the manual backward walk. I'll rework it to use a forward matcher from dot through the attention ops to the convert that find_softmax_base_ops inserts and adjust the ordering, so it runs after find_softmax_base_ops.

Would it be possible to merge the current version once CI passes, and I'll push the composed matcher refactor as a follow-up next week? The end result is functionally identical, same graph transformation, same test results, just a better use of the matcher framework. This is a high priority item.

**Comment by aditya-dl** [comment]
@causten can we restart the CI/CD?

**Comment by aditya-dl** [comment]
> ## [Codecov](https://app.codecov.io/gh/ROCm/AMDMIGraphX/pull/4677?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ROCm) Report
> ❌ Patch coverage is `98.21429%` with `1 line` in your changes missing coverage. Please review.
> 
> [Files with missing lines](https://app.codecov.io/gh/ROCm/AMDMIGraphX/pull/4677?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ROCm)	Patch %	Lines
> [src/rewrite_reduce.cpp](https://app.codecov.io/gh/ROCm/AMDMIGraphX/pull/4677?src=pr&el=tree&filepath=src%2Frewrite_reduce.cpp&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ROCm#diff-c3JjL3Jld3JpdGVfcmVkdWNlLmNwcA==)	98.04%	[1 Missing ⚠️ ](https://app.codecov.io/gh/ROCm/AMDMIGraphX/pull/4677?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ROCm)
> Additional details and impacted files
> ```diff
> @@             Coverage Diff             @@
> ##           develop    #4677      +/-   ##
> ===========================================
> + Coverage    92.43%   92.49%   +0.06%     
> ===========================================
>   Files          583      583              
>   Lines        29251    29566     +315     
> ===========================================
> + Hits         27037    27347     +310     
> - Misses        2214     2219       +5     
> ```
> 
> [Files with missing lines](https://app.codecov.io/gh/ROCm/AMDMIGraphX/pull/4677?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ROCm)	Coverage Δ	
> [src/fuse_attention.cpp](https://app.codecov.io/gh/ROCm/AMDMIGraphX/pull/4677?src=pr&el=tree&filepath=src%2Ffuse_attention.cpp&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ROCm#diff-c3JjL2Z1c2VfYXR0ZW50aW9uLmNwcA==)	`98.00% <100.00%> (+<0.01%)`	⬆️
> [src/rewrite_reduce.cpp](https://app.codecov.io/gh/ROCm/AMDMIGraphX/pull/4677?src=pr&el=tree&filepath=src%2Frewrite_reduce.cpp&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ROCm#diff-c3JjL3Jld3JpdGVfcmVkdWNlLmNwcA==)	`99.38% <98.04%> (-0.62%)`	⬇️
> ... and [18 files with indirect coverage changes](https://app.codecov.io/gh/ROCm/AMDMIGraphX/pull/4677/indirect-changes?src=pr&el=tree-more&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=ROCm)
> 
> 🚀 New features to boost your workflow:

The uncovered line is `return std::nullopt` inside the step lambda when `name == "dot"`. This is unreachable because `std::find_if` returns as soon as it finds the dot element; it never calls `step(dot)` to advance past it. The line exists for unfold's termination contract but `find_if` short-circuits before it's executed.

File: src/rewrite_reduce.cpp [hunk 8]
```diff
@@ -79,6 +80,94 @@ struct find_softmax
...
+            {
+                auto i0 = current->inputs()[0];
+                auto i1 = current->inputs()[1];
+                if(i0->name() == "dot" or i0->name() == "where" or i0->name() == "convert" or
+                   i0->name() == "mul")
+                    return i0;
+                return i1;
+            }
+            if(name == "broadcast" or name == "multibroadcast" or name == "convert")
+                return current->inputs()[0];
```
Comment (pfultz2) [inline src/rewrite_reduce.cpp]: Checking for operator names is brittle. I think this can be simplified to:


```suggestion
            if(current->inputs().size() == 1)
                return current->inputs().front();
            auto it = std::find_if(current->inputs().begin(), current->inputs().end(), [](instruction_ref input) {
                return not input->can_eval();
            });
            if(it == current->inputs().end())
                return std::nullopt;
            return *it;
```

File: src/rewrite_reduce.cpp (hunk 7 above)
Comment (pfultz2) [inline src/rewrite_reduce.cpp:307]: This should run after `find_softmax_base_ops`:


```suggestion
        match::find_matches(m, find_softmax_base_ops{});
        match::find_matches(m, find_dot_softmax_fp32{});
```

File: src/rewrite_reduce.cpp (hunk 6 above)
Comment (pfultz2) [reply src/rewrite_reduce.cpp]: I think its better to walk up even in the matcher, but we need the `skip` matcher to traverse the non-constant paths and not just unary ops. Its probably fine to traverse up in `apply` for now. Just need some minor updates to simplify it.

File: src/rewrite_reduce.cpp (hunk 8 above)
Comment (aditya-dl) [reply src/rewrite_reduce.cpp]: Implemented the `can_eval()` approach: replaced the name-based input checking with a generic rule; follow the non-evaluable, non-bool input at each step. This handles any op on the data path without hardcoding op names.

One addition to the suggestion: for `where(mask, -inf, scores)`, both the mask and scores are non-evaluable (mask depends on the sequence length parameter). Added a `bool_type` check to skip the mask condition, so `find_if` correctly follows the attention data path (scores) instead of the mask.
 ```
 auto it = std::find_if(
      current->inputs().begin(),
      current->inputs().end(),
      [](instruction_ref input) {
          return not input->can_eval() and
                 input->get_shape().type() != shape::bool_type;
      });
```
Updated tests accordingly — relu is now traversed (single-input op), scale_left uses a literal scale matching real model patterns.

File: src/rewrite_reduce.cpp (hunk 7 above)
Comment (aditya-dl) [reply src/rewrite_reduce.cpp:307]: Running `find_dot_softmax_fp32` after `find_softmax_base_ops` results in the failure of unit tests. `match::softmax()` can't match after `find_softmax_base_ops` wraps each `softmax` op in convert pairs. `skip_broadcasts` in the softmax matcher skips broadcasts but not converts.

It works if we add eliminate_convert between the two matchers (to collapse the convert pairs first), but that means a second eliminate_convert + DCE + CSE pass:
```
  match::find_matches(m, find_softmax_base_ops{});
  run_passes(m, {eliminate_convert{}, DCE{}, CSE{}});
  match::find_matches(m, find_dot_softmax_fp32{});
  run_passes(m, {eliminate_convert{}, DCE{}, CSE{}});
```
Currently it runs before with a single pass at the end. Both orderings produce the same graph. Would you prefer the reordered version with the extra pass, or is the current ordering acceptable?

**Review by pfultz2 — APPROVED** [review]

**Comment by aditya-dl** [comment]
updated formatting

**Review by shivadbhavsar — APPROVED** [review]
